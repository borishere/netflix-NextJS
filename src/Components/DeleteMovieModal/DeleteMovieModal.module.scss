.delete-movie-modal {
  display: block;

  .modal-content {
    max-width: 566px;
  }
}