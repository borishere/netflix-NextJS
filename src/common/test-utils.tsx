import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ModalContext } from '../Context/ModalContext';
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import { createMemoryHistory } from 'history';
import { moviesApi } from '../services/movies';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import moviesReducer from '../app/moviesSlice';
import { configureStore } from '@reduxjs/toolkit';

const history = createMemoryHistory();


const AllTheProviders: FC = ({ children }) => {
  return (
    <Router location='/' navigator={history}>
      <Provider store={store}>
        <ModalContext.Provider value={{}}>
          {children}
        </ModalContext.Provider>
      </Provider>
    </Router>
  );
};

function renderWithStore(
  ui: ReactElement,
  {
    store = configureStore({
      reducer: {
        [moviesApi.reducerPath]: moviesApi.reducer,
        movies: moviesReducer

      },
      middleware: (getDefaultMiddleware: any) =>
        getDefaultMiddleware().concat(moviesApi.middleware)
    }),
    ...renderOptions
  } = {}
) {
  setupListeners(store.dispatch);

  function Wrapper({ children }: any) {
    const history = createMemoryHistory();

    return (
      <Provider store={store}>
        <Router location='/' navigator={history}>
          {children}
        </Router>
      </Provider>
    );
  }

  store.dispatch(moviesApi.util.resetApiState());

  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

const root = document.createElement('div');
root.id = 'root';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, container: document.body.appendChild(root), ...options });

export * from '@testing-library/react';
export { customRender, renderWithStore };