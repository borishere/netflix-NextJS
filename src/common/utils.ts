export function addParamToExistsSearchParams(key: string, value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.set(key, value);

  return params;
}

export function deleteParamFromExistsSearchParams(value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.delete(value);

  return params;
}

export const BASE_API_URL: string = 'http://localhost:4000/';