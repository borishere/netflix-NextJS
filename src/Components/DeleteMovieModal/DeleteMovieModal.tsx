import React, { FC } from 'react';
import { useDeleteMovieMutation } from '../../services/movies';
import { Modal, ModalProps } from '../Modal/Modal';
import './DeleteMovieModal.scss';

interface Props extends ModalProps {
  movieId: number;
}

export const DeleteMovieModal: FC<Props> = ({ movieId, ...props }) => {
  const [deleteMovie, { isLoading }] = useDeleteMovieMutation();

  const onSubmit = (): void => {
    if (movieId) {
      deleteMovie(movieId).then(() => {
        props.show(false);
      });
    }
  };

  if (!movieId) {
    return null;
  }

  return (
    <Modal modalClass='delete-movie-modal' {...props}>
      <>
        {isLoading && <span>Deleting...</span>}
        <h2 className='modal-title'>DELETE MOVIE</h2>
        <p>Are you sure you want to delete this movie?</p>
        <div className='modal-buttons'>
          <button className='submit-btn' onClick={onSubmit}>CONFIRM</button>
        </div>
      </>
    </Modal>
  );
};