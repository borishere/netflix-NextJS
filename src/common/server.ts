import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { BASE_API_URL } from './utils';
import mockData from './__mocks__/movies.json';

let movies: any[] = Array.from(mockData);

export const server = setupServer(
  rest.get(`${BASE_API_URL}movies`, (req, res, ctx) => {
    return res(ctx.json(movies.slice(0, 2)));
  }),

  rest.get(`${BASE_API_URL}movies:id`, (req, res, ctx) => {
    const { id } = req.params;
    const movie = movies.find((movie) => movie.id === id);

    return res(ctx.json(movie));
  }),

  rest.post(`${BASE_API_URL}movies`, (req, res, ctx) => {
    movies.push(req.body);

    return res(ctx.status(200));
  }),

  rest.put(`${BASE_API_URL}movies`, (req, res, ctx) => {
    const requestMovie: any = req.body;

    const movieIndex = movies.findIndex((m) => m.id === requestMovie.id);

    if (movieIndex < 0) {
      return res(ctx.status(404));
    }

    movies[movieIndex] = requestMovie;

    return res(ctx.status(200));
  }),

  rest.delete(`${BASE_API_URL}movies/:id`, (req, res, ctx) => {
    const { id } = req.params;

    movies = movies.filter((m) => m.id !== id);

    return res(ctx.status(200));
  })
);
