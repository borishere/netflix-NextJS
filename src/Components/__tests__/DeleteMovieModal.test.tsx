import React from 'react';
import { screen, customRender, waitFor } from '../../common/test-utils';
import { DeleteMovieModal } from '../DeleteMovieModal/DeleteMovieModal';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

describe('DeleteMovieModal', () => {
  test('is visible', () => {
    const { asFragment } = customRender(
      <DeleteMovieModal
        isShown={true}
        show={jest.fn()}
        movieId={1}
      />
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('DELETE MOVIE')).toBeInTheDocument();
  });

  test('is hidden', () => {
    customRender(
      <DeleteMovieModal
        isShown={false}
        show={jest.fn()}
        movieId={1}
      />
    );

    expect(screen.queryByText('DELETE MOVIE')).not.toBeInTheDocument();
  });

  test('Close on submit', async () => {
    const mockShow = jest.fn();

    customRender(
      <DeleteMovieModal
        isShown={true}
        show={mockShow}
        movieId={1}
      />
    );

    userEvent.click(screen.getByRole('button', { name: /confirm/i }));

    await waitFor(() => {
      expect(mockShow).toHaveBeenCalledTimes(1);
    });
  });
});