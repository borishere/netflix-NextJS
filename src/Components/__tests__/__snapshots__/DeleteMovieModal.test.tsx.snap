// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteMovieModal is visible 1`] = `
<DocumentFragment>
  <div
    class="modal delete-movie-modal"
  >
    <div
      class="modal-content"
    >
      <img
        aria-label="close"
        class="modal-close-btn"
        src="test-file-stub"
      />
      <div
        class="modal-body"
      >
        <h2
          class="modal-title"
        >
          DELETE MOVIE
        </h2>
        <p>
          Are you sure you want to delete this movie?
        </p>
        <div
          class="modal-buttons"
        >
          <button
            class="submit-btn"
          >
            CONFIRM
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
